Fix missing preprocessor definition in build.ninja

radoslawcybulski opened the pull request:
Add missing preprocessor definition to build.ninja (in configure.py), which enables precompiled headers, when non-cmake mode is used.

`stdafx.hh.pch` generated file is removed from ccache - when editing source files while compiling the compiler might generate pch file, which no longer will match timestamps / sizes of header files. If the mismatched pch file ends in ccache cache it will have to be removed manually or `stdafx.hh` will have to be changed.

nyh reviewed (changes requested):
I'm afraid I don't understand what I'm seeing in this patch:
1. It changes something in Seastar submodule, is this by mistake?
2. It adds a -Dsomething in configure.py - this is mentioned in the commit message, but to be honest - I didn't understand the explanation.
3. It makes stdafs.hh.pcf not cached by ccache. Which is also mentioned in the same paragraph of where you explained the -D, but I don't understand how the two changes are related.
4. You modified stdafx.hh and removed a bunch of includes from it. Why? Were they wrong? Or what? This is not mentioned in the commit message. As seems to be the second (or third) unrelated thing you're doing in this patch, please consider splitting it to multiple patches, or the very least, split the commit message to separate paragraphs explaining the different changes.
5. You modified test/perf/perf_stable.cc - I have no idea what's that about or how it's related to this PR.

radoslawcybulski commented:
> I'm afraid I don't understand what I'm seeing in this patch:
> 
>     1. It changes something in Seastar submodule, is this by mistake?
> 
>     2. It adds a -Dsomething in configure.py - this is mentioned in the commit message, but to be honest - I didn't understand the explanation.
> 
>     3. It makes stdafs.hh.pcf not cached by ccache. Which is also mentioned in the same paragraph of where you explained the -D, but I don't understand how the two changes are related.
> 
>     4. You modified stdafx.hh and removed a bunch of includes from it. Why? Were they wrong? Or what? This is not mentioned in the commit message. As seems to be the second (or third) unrelated thing you're doing in this patch, please consider splitting it to multiple patches, or the very least, split the commit message to separate paragraphs explaining the different changes.
> 
>     5. You modified test/perf/perf_stable.cc - I have no idea what's that about or how it's related to this PR.

1. No, this is exposed by "double include" (precompiled header is essentially all includes done twice, the second one should be noop due to include guards, but if those are missing, then we get errors)
2. Without this flag precompiled header will be empty - this is so you can fast compile, if you due a lot of changes in seastar headers and compile scylla (otherwise changing a thing in header will recompile whole scylla every single time)
3. i got a situation once, where i got pch file with mismatched timestamp (because i edited header during compilation, so ccache stored precompiled header with old timestamp info and now i got new one) and it turns out compiler checks this later on and complains. And now i've no way of getting old timestamp for header (because i don't know it) and i can't easily remove pch from ccache's cache (because ccache doesn't allow to remove single file, you can only purge all). Since pch is compiled not that often i decided to remove it from ccache for safety.
4. I compiled it against newest seastar master and removed headers, that are gone on seastar repo. Test headers show another issue - multiple instances of `main` function (because some headers do define this one).
5. This file defines two boost macros, which previously were undefined and now they are, because pch file pulls them in.

I overestimated simplicity of build.ninja format (just found out another issue with build order), unfortunately. I will revert my original patch, this one needs more work.

radoslawcybulski commented:
My idea was to run fix quickly - as you know by now as well it's not that simple. I'm making revert PR as we speak, will release seastar thing first (as it's simplest) and then others, possibly in separate PRs.

radoslawcybulski commented:
> Years of neglect (while we were always "switch to cmake in a few days") made configure.py a big mess - today I want to fix #17361 and underestimated how hairy it will be to fix it.

I did some timing and ninja.build from configure.py is faster by approximately 10-20%. No idea why, i've not yet parsed command logs. So while it's messy it has its use.
